In [1]:
# Import necessary libraries
import pandas as pd
from sklearn.datasets import load_wine

# Load wine dataset from sklearn library
wine_data = load_wine()

# Create first dataframe with wine information 
wine_info_df = pd.DataFrame(data=wine_data.data, columns=wine_data.feature_names) 

# Create second dataframe with target information (will be used and explained later)
wine_target_df = pd.DataFrame({'target': wine_data.target})

#Let's look at our dataframe
wine_info_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols   
0      14.23        1.71  2.43               15.6      127.0           2.80  \
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue   
0          3.06                  0.28             2.29             5.64  1.04  \
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

We can read that we have 178 rows and 13 columns.

In [4]:
# If you don't want to use sklern library, you can download the same dataset from github
# Use read_csv() function to create dataframe
wine_csv_df = pd.read_csv("Wine.csv")
wine_csv_df

Wine  Alcohol  Malic.acid   Ash   Acl   Mg  Phenols  Flavanoids   
0       1    14.23        1.71  2.43  15.6  127     2.80        3.06  \
1       1    13.20        1.78  2.14  11.2  100     2.65        2.76   
2       1    13.16        2.36  2.67  18.6  101     2.80        3.24   
3       1    14.37        1.95  2.50  16.8  113     3.85        3.49   
4       1    13.24        2.59  2.87  21.0  118     2.80        2.69   
..    ...      ...         ...   ...   ...  ...      ...         ...   
173     3    13.71        5.65  2.45  20.5   95     1.68        0.61   
174     3    13.40        3.91  2.48  23.0  102     1.80        0.75   
175     3    13.27        4.28  2.26  20.0  120     1.59        0.69   
176     3    13.17        2.59  2.37  20.0  120     1.65        0.68   
177     3    14.13        4.10  2.74  24.5   96     2.05        0.76   

     Nonflavanoid.phenols  Proanth  Color.int   Hue    OD  Proline  
0                    0.28     2.29       5.64  1.04  3.92     1065  
1                    0.26     1.28       4.38  1.05  3.40     1050  
2                    0.30     2.81       5.68  1.03  3.17     1185  
3                    0.24     2.18       7.80  0.86  3.45     1480  
4                    0.39     1.82       4.32  1.04  2.93      735  
..                    ...      ...        ...   ...   ...      ...  
173                  0.52     1.06       7.70  0.64  1.74      740  
174                  0.43     1.41       7.30  0.70  1.56      750  
175                  0.43     1.35      10.20  0.59  1.56      835  
176                  0.53     1.46       9.30  0.60  1.62      840  
177                  0.56     1.35       9.20  0.61  1.60      560  

[178 rows x 14 columns]

In [6]:
# Look at few first (default=5) rows of the data
wine_info_df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols   
0    14.23        1.71  2.43               15.6      127.0           2.80  \
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue   
0        3.06                  0.28             2.29             5.64  1.04  \
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [7]:
# In head() and tail() method you can increase the number of rows by passing the number as an argument
# Last 10 rows of our DataFrame
wine_info_df.tail(10)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols   
168    13.58        2.58  2.69               24.5      105.0           1.55  \
169    13.40        4.60  2.86               25.0      112.0           1.98   
170    12.20        3.03  2.32               19.0       96.0           1.25   
171    12.77        2.39  2.28               19.5       86.0           1.39   
172    14.16        2.51  2.48               20.0       91.0           1.68   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue   
168        0.84                  0.39             1.54         8.660000  0.74  \
169        0.96                  0.27             1.11         8.500000  0.67   
170        0.49                  0.40             0.73         5.500000  0.66   
171        0.51                  0.48             0.64         9.899999  0.57   
172        0.70                  0.44             1.24         9.700000  0.62   
173        0.61                  0.52             1.06         7.700000  0.64   
174        0.75                  0.43             1.41         7.300000  0.70   
175        0.69                  0.43             1.35        10.200000  0.59   
176        0.68                  0.53             1.46         9.300000  0.60   
177        0.76                  0.56             1.35         9.200000  0.61   

     od280/od315_of_diluted_wines  proline  
168                          1.80    750.0  
169                          1.92    630.0  
170                          1.83    510.0  
171                          1.63    470.0  
172                          1.71    660.0  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0

In [8]:
# Information about our dataset: how many entries, number of column, column names, number of non-null values
# and type of each column, moreover we can read what class is it and how much memory it uses
wine_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
dtypes: fl

In [9]:
# Getting descriptive statistics about the dataset
wine_info_df.describe()

alcohol  malic_acid         ash  alcalinity_of_ash   magnesium   
count  178.000000  178.000000  178.000000         178.000000  178.000000  \
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins   
count     178.000000  178.000000            178.000000       178.000000  \
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       color_intensity         hue  od280/od315_of_diluted_wines      proline  
count       178.000000  178.000000                    178.000000   178.000000  
mean          5.058090    0.957449                      2.611685   746.893258  
std           2.318286    0.228572                      0.709990   314.907474  
min           1.280000    0.480000                      1.270000   278.000000  
25%           3.220000    0.782500                      1.937500   500.500000  
50%           4.690000    0.965000                      2.780000   673.500000  
75%           6.200000    1.120000                      3.170000   985.000000  
max          13.000000    1.710000                      4.000000  1680.000000

In [10]:
# We already know what are the dimensions of our dataset, but we can always retrieve this value:
wine_info_df.shape

(178, 13)

In [11]:
# Using loc to localise all rows within the 'alcohol' and 'proline' column
wine_info_df.loc[:, ['alcohol', 'proline']]

alcohol  proline
0      14.23   1065.0
1      13.20   1050.0
2      13.16   1185.0
3      14.37   1480.0
4      13.24    735.0
..       ...      ...
173    13.71    740.0
174    13.40    750.0
175    13.27    835.0
176    13.17    840.0
177    14.13    560.0

[178 rows x 2 columns]

In [12]:
# Using iloc to localise all rows from 1st to 3rd column (Index starts from 0)
wine_info_df.iloc[:, 0:3]

alcohol  malic_acid   ash
0      14.23        1.71  2.43
1      13.20        1.78  2.14
2      13.16        2.36  2.67
3      14.37        1.95  2.50
4      13.24        2.59  2.87
..       ...         ...   ...
173    13.71        5.65  2.45
174    13.40        3.91  2.48
175    13.27        4.28  2.26
176    13.17        2.59  2.37
177    14.13        4.10  2.74

[178 rows x 3 columns]

In [14]:
# Using iloc to localise first 10 rows from 1st to 4rd column (Index starts from 0)
wine_info_df.iloc[0:10, 0:4]

alcohol  malic_acid   ash  alcalinity_of_ash
0    14.23        1.71  2.43               15.6
1    13.20        1.78  2.14               11.2
2    13.16        2.36  2.67               18.6
3    14.37        1.95  2.50               16.8
4    13.24        2.59  2.87               21.0
5    14.20        1.76  2.45               15.2
6    14.39        1.87  2.45               14.6
7    14.06        2.15  2.61               17.6
8    14.83        1.64  2.17               14.0
9    13.86        1.35  2.27               16.0

In [16]:
# Check what are the names of columns in the dataset
print(wine_info_df.columns)

# Drop 'ash' column and create new DataFrame with the same name as before
wine_info_df = wine_info_df.drop(columns=['ash'])

# Chech whether the column is still in the dataset
print('ash' in wine_info_df.columns)

Index(['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium',
       'total_phenols', 'flavanoids', 'nonflavanoid_phenols',
       'proanthocyanins', 'color_intensity', 'hue',
       'od280/od315_of_diluted_wines', 'proline'],
      dtype='object')
False


In [17]:
# Drop 'malic_acid'
wine_info_df.drop(columns=['malic_acid'])
# Data will be displayed without `malic_acid' column

alcohol  alcalinity_of_ash  magnesium  total_phenols  flavanoids   
0      14.23               15.6      127.0           2.80        3.06  \
1      13.20               11.2      100.0           2.65        2.76   
2      13.16               18.6      101.0           2.80        3.24   
3      14.37               16.8      113.0           3.85        3.49   
4      13.24               21.0      118.0           2.80        2.69   
..       ...                ...        ...            ...         ...   
173    13.71               20.5       95.0           1.68        0.61   
174    13.40               23.0      102.0           1.80        0.75   
175    13.27               20.0      120.0           1.59        0.69   
176    13.17               20.0      120.0           1.65        0.68   
177    14.13               24.5       96.0           2.05        0.76   

     nonflavanoid_phenols  proanthocyanins  color_intensity   hue   
0                    0.28             2.29             5.64  1.04  \
1                    0.26             1.28             4.38  1.05   
2                    0.30             2.81             5.68  1.03   
3                    0.24             2.18             7.80  0.86   
4                    0.39             1.82             4.32  1.04   
..                    ...              ...              ...   ...   
173                  0.52             1.06             7.70  0.64   
174                  0.43             1.41             7.30  0.70   
175                  0.43             1.35            10.20  0.59   
176                  0.53             1.46             9.30  0.60   
177                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 11 columns]

In [18]:
# Chech whether the 'malic_acid' column is still in the dataset
print('malic_acid' in wine_info_df.columns)

True


In [19]:
# To avoid that, we can just use `inplace=True`
wine_info_df.drop(columns=['malic_acid'], inplace=True)

In [20]:
# Chech whether the 'malic_acid' column is still in the dataset
print('malic_acid' in wine_info_df.columns)
# Finally it worked :) 

False


In [21]:
# Renaming column to add more context
wine_info_df = wine_info_df.rename(columns={'alcohol': 'alcohol [%]'})

In [22]:
# Merge two DataFrames together
merged_df = pd.merge(wine_info_df, wine_target_df, left_index=True, right_index=True)
merged_df

alcohol [%]  alcalinity_of_ash  magnesium  total_phenols  flavanoids   
0          14.23               15.6      127.0           2.80        3.06  \
1          13.20               11.2      100.0           2.65        2.76   
2          13.16               18.6      101.0           2.80        3.24   
3          14.37               16.8      113.0           3.85        3.49   
4          13.24               21.0      118.0           2.80        2.69   
..           ...                ...        ...            ...         ...   
173        13.71               20.5       95.0           1.68        0.61   
174        13.40               23.0      102.0           1.80        0.75   
175        13.27               20.0      120.0           1.59        0.69   
176        13.17               20.0      120.0           1.65        0.68   
177        14.13               24.5       96.0           2.05        0.76   

     nonflavanoid_phenols  proanthocyanins  color_intensity   hue   
0                    0.28             2.29             5.64  1.04  \
1                    0.26             1.28             4.38  1.05   
2                    0.30             2.81             5.68  1.03   
3                    0.24             2.18             7.80  0.86   
4                    0.39             1.82             4.32  1.04   
..                    ...              ...              ...   ...   
173                  0.52             1.06             7.70  0.64   
174                  0.43             1.41             7.30  0.70   
175                  0.43             1.35            10.20  0.59   
176                  0.53             1.46             9.30  0.60   
177                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  target  
0                            3.92   1065.0       0  
1                            3.40   1050.0       0  
2                            3.17   1185.0       0  
3                            3.45   1480.0       0  
4                            2.93    735.0       0  
..                            ...      ...     ...  
173                          1.74    740.0       2  
174                          1.56    750.0       2  
175                          1.56    835.0       2  
176                          1.62    840.0       2  
177                          1.60    560.0       2  

[178 rows x 12 columns]

We can see new column 'target'

In [24]:
# Concatenate DataFrames along columns, if dimensions are not the same, missing values will be replaced with 'NaN'
additional_info_df = pd.DataFrame({'additional_info': ['Info_A', 'Info_B', 'Info_C']})
concatenated_df = pd.concat([wine_info_df, additional_info_df], axis=1)
concatenated_df

alcohol [%]  alcalinity_of_ash  magnesium  total_phenols  flavanoids   
0          14.23               15.6      127.0           2.80        3.06  \
1          13.20               11.2      100.0           2.65        2.76   
2          13.16               18.6      101.0           2.80        3.24   
3          14.37               16.8      113.0           3.85        3.49   
4          13.24               21.0      118.0           2.80        2.69   
..           ...                ...        ...            ...         ...   
173        13.71               20.5       95.0           1.68        0.61   
174        13.40               23.0      102.0           1.80        0.75   
175        13.27               20.0      120.0           1.59        0.69   
176        13.17               20.0      120.0           1.65        0.68   
177        14.13               24.5       96.0           2.05        0.76   

     nonflavanoid_phenols  proanthocyanins  color_intensity   hue   
0                    0.28             2.29             5.64  1.04  \
1                    0.26             1.28             4.38  1.05   
2                    0.30             2.81             5.68  1.03   
3                    0.24             2.18             7.80  0.86   
4                    0.39             1.82             4.32  1.04   
..                    ...              ...              ...   ...   
173                  0.52             1.06             7.70  0.64   
174                  0.43             1.41             7.30  0.70   
175                  0.43             1.35            10.20  0.59   
176                  0.53             1.46             9.30  0.60   
177                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline additional_info  
0                            3.92   1065.0          Info_A  
1                            3.40   1050.0          Info_B  
2                            3.17   1185.0          Info_C  
3                            3.45   1480.0             NaN  
4                            2.93    735.0             NaN  
..                            ...      ...             ...  
173                          1.74    740.0             NaN  
174                          1.56    750.0             NaN  
175                          1.56    835.0             NaN  
176                          1.62    840.0             NaN  
177                          1.60    560.0             NaN  

[178 rows x 12 columns]

In [26]:
# Group data by target and aggregate alcohol by its average for given target and maximum value for proline
grouped_data = merged_df.groupby('target').agg({'alcohol [%]': 'mean', 'proline': 'max'})
grouped_data

alcohol [%]  proline
target                      
0         13.744746   1680.0
1         12.278732    985.0
2         13.153750    880.0